change uefi packed structs to new integer backed syntax

truemedian opened the pull request:
* device_path_protocol now uses extern structs with align(1) fields because
the transition to integer backed packed struct broke alignment

* added comptime asserts that device_path_protocol structs do not violate
alignment and size specifications

ominitay commented on lib/std/os/uefi/protocols/device_path_protocol.zig:
```diff
@@ -469,7 +484,7 @@ pub const MessagingDevicePath = union(Subtype) {
     };
 
-    pub const UartDevicePath = packed struct {
+    pub const UartDevicePath = extern struct {
```
There is a similar alignment problem here, with 2 bytes of padding between `reserved` and `baud_rate`.

ominitay reviewed (changes requested):
You need to check over each change from `packed` -> `extern` struct and ensure that field alignment is correct. I've provided two examples for you in this review.

ominitay commented on lib/std/os/uefi/protocols/device_path_protocol.zig:
```diff
@@ -491,19 +799,43 @@ pub const MessagingDevicePath = union(Subtype) {
         length: u16,
-        reserved: u16,
-        baud_rate: u32,
+        reserved: u32,
+        baud_rate: u64,
```
This doesn't fix the problem. You've just made that accidental padding explicit. Use a similar solution as the above.

truemedian replied:
This is actually correct, UEFI defines `reserved` as having 4 bytes and `baud_rate` as being 8 bytes. I'm not sure why these were 2 and 4. Check `Table 10-34 UART Device Path`

ominitay replied:
Fair enough, was comparing against the original (and assuming it was correct).

ominitay commented on lib/std/os/uefi/protocols/device_path_protocol.zig:
```diff
@@ -169,27 +202,70 @@ pub const HardwareDevicePath = union(Subtype) {
         length: u16,
         interface_type: u8,
-        base_address: usize,
+        base_address: u64 align(1),
```
This is the right way to go!

ominitay reviewed (changes requested):
I haven't been able to check everything over yet. I'll give each struct a closer look to check the alignments tomorrow :)

andrewrk commented:
Thanks for the help reviewing, @ominitay.

The CI failure will be fixed by running `zig fmt`:
```
Looking for non-conforming code formatting...
../lib/std/os/uefi/protocols/device_path_protocol.zig
````

fifty-six commented:
Might be worth adding `std.testing.refAllDeclsRecursive(@This())` to uefi.zig to ensure that all the packed struct stuff gets checked and stays fixed fwiw

truemedian commented:
The issue with running tests on them is that there is no testing infrastructure for uefi, there is simply no way to run them without doing something weird with the test runner. And you can't safely run the tests on another platform because of potential differences in the abi

fifty-six commented:
Ah ok. I assume that would change extern struct stuff then? Because refAllDecls would just be to ensure it stays compiling, not much else afaik.

ifreund commented on lib/std/os/uefi/protocols/absolute_pointer_protocol.zig:
```diff
@@ -31,27 +31,29 @@ pub const AbsolutePointerProtocol = extern struct {
     };
 };
 
+pub const AbsolutePointerModeAttributes = packed struct(u32) {
+    supports_alt_active: bool,
+    supports_pressure_as_z: bool,
+};
```
This won't compile, padding must be explicit even with the new semantics. The difference is that the compiler will check that the total size/alignment matches that of a u32

truemedian replied:
Yeah, I just now realized this, fixing it as we speak

truemedian commented on lib/std/os/uefi/protocols/device_path_protocol.zig:
```diff
@@ -4,7 +4,7 @@ const uefi = std.os.uefi;
 const Guid = uefi.Guid;
 
-pub const DevicePathProtocol = packed struct {
+pub const DevicePathProtocol = extern struct {
```
You're right, I think this might also apply everywhere else as well in device_path_protocol

truemedian commented:
It would appear that the change making packed structs backed by integers also made them no longer align(1). So I guess I also need to go through and align(1) all the fields manually.

fifty-six commented on lib/std/os/uefi/protocols/device_path_protocol.zig:
```diff
@@ -49,14 +52,20 @@ pub const DevicePathProtocol = packed struct {
-        var ptr = @ptrCast([*:0]u16, @alignCast(2, @ptrCast([*]u8, new)) + @sizeOf(MediaDevicePath.FilePathDevicePath));
+        var ptr = @ptrCast(
+            [*:0]u16,
+            @alignCast(2, @ptrCast([*]u8, new) + @sizeOf(MediaDevicePath.FilePathDevicePath)),
```
Probably don't want the `@alignCast`s here now

truemedian replied:
That needs to be aligned because u16 is align(2)

fifty-six replied:
Oh sorry I meant the one after that with `DevicePathProtocol`, I believe the u16 strings are aligned. `FilePathDevicePath` shouldn't be aligned either I think, given it's a subtype of `DevicePathProtocol`.

fifty-six commented on lib/std/os/uefi/protocols/device_path_protocol.zig:
```diff
@@ -100,14 +106,26 @@ pub const DevicePathProtocol = packed struct {
+                return @unionInit(TUnion, subtype.name, @ptrCast(
+                    subtype.field_type,
+                    @alignCast(@alignOf(subtype.field_type), self),
+                ));
```
This alignCast shouldn't be here as its a pointer to (a subtype of) DevicePathProtocol which is unaligned.

truemedian replied:
good catch, totally forgot about this

fifty-six commented on lib/std/os/uefi/protocols/device_path_protocol.zig:
```diff
@@ -23,7 +23,10 @@ pub const DevicePathProtocol = packed struct {
-        return @ptrCast(*DevicePathProtocol, @ptrCast([*]u8, self) + self.length);
+        return @ptrCast(
+            *DevicePathProtocol,
+            @alignCast(2, @ptrCast([*]u8, self) + self.length),
```
This also shouldn't be aligned

fifty-six commented:
Tested this with my bootloader and it works again after changing those 2 alignCasts! :tada:

truemedian commented:
I've cleaned up the commit history, if people can verify that it also works for them it seems like this can be merged.

fifty-six commented:
Can confirm it still works, lgtm

ominitay commented on lib/std/os/uefi/protocols/device_path_protocol.zig:
```diff
@@ -292,130 +402,288 @@ pub const MessagingDevicePath = union(Subtype) {
         subtype: Subtype,
-        length: u16,
-        interface_number: u16,
-        device_vendor_id: u16,
-        device_product_id: u16,
+        length: u16 align(1),
+        interface_number: u16 align(1),
+        device_vendor_id: u16 align(1),
+        device_product_id: u16 align(1),
 
         pub fn serial_number(self: *const UsbWwidDevicePath) []const u16 {
             var serial_len = (self.length - @sizeOf(UsbWwidDevicePath)) / @sizeOf(u16);
-            return @ptrCast([*]u16, @ptrCast([*]u8, self) + @sizeOf(UsbWwidDevicePath))[0..serial_len];
+            return @ptrCast(
+                [*]const u16,
+                @alignCast(2, @ptrCast([*]const u8, self) + @sizeOf(UsbWwidDevicePath)),
+            )[0..serial_len];
         }
     };
```
Same question about alignment here (and in a few other places).

truemedian replied:
DevicePaths are not aligned in memory, but strings must be

ominitay replied:
AlignCast doesn't change the alignment of this though. You should have an under-aligned pointer instead, surely.

truemedian replied:
Given I can't find anything about how this string is aligned and that it directly follows this struct, I think its correct to underalign this

ominitay reviewed (changes requested):
Sorry this review took a while to get you to you :)

truemedian commented on lib/std/os/uefi/protocols/hii.zig:
```diff
@@ -4,7 +4,7 @@ const Guid = uefi.Guid;
 
 /// The header found at the start of each package.
-pub const HIIPackageHeader = packed struct {
+pub const HIIPackageHeader = packed struct(u32) {
```
It needs to be packed because of the `u24`

truemedian replied:
Nope, u24 isn't defined by the ABI

truemedian commented on lib/std/os/uefi/protocols/edid_override_protocol.zig:
```diff
@@ -31,8 +29,8 @@ pub const EdidOverrideProtocol = extern struct {
     dont_override: bool,
     enable_hot_plug: bool,
-    _pad: u30 = 0,
+    _pad: u30,
```
Part of these changes I merged in from another branch, I had removed them because most of the structs come from uefi. I can see how it would help with initializing them though.

truemedian commented on lib/std/os/uefi/protocols/device_path_protocol.zig:
```diff
@@ -204,37 +280,71 @@ pub const AcpiDevicePath = union(Subtype) {
+    pub const AdrDevicePath = extern struct {
         type: DevicePathType,
         subtype: Subtype,
-        length: u16,
-        adr: u32,
+        length: u16 align(1),
+        adr: u32 align(1),
+
         // multiple adr entries can optionally follow
         pub fn adrs(self: *const AdrDevicePath) []const u32 {
             // self.length is a minimum of 8 with one adr which is size 4.
             var entries = (self.length - 4) / @sizeOf(u32);
-            return @ptrCast([*]const u32, &self.adr)[0..entries];
+            return @ptrCast([*]const u32, @alignCast(4, &self.adr))[0..entries];
```
The struct declaration is align(1) because all DevicePaths are laid out in memory right after eachother with no padding. I'm actually leaning towards adrs being align(1) as well because the spec only says they are 4 bytes, not that they are u32

ominitay replied:
If the struct is meant to be align(1), then the alignCast is wrong.

truemedian replied:
Yeah, I just changed this to return `[]align(1) const u32`, given I'm pretty sure that UEFI doesn't give us any alignment guarantees here

truemedian commented on lib/std/os/uefi/protocols/device_path_protocol.zig:
```diff
@@ -292,130 +402,288 @@ pub const MessagingDevicePath = union(Subtype) {
         type: DevicePathType,
         subtype: Subtype,
-        length: u16,
-        local_ip_address: uefi.Ipv4Address,
-        remote_ip_address: uefi.Ipv4Address,
```
Whoops, this is a artifact of trying to use packed structs

ominitay commented on lib/std/os/uefi/protocols/device_path_protocol.zig:
```diff
@@ -545,81 +875,184 @@ pub const MediaDevicePath = union(Subtype) {
+
+    pub const FilePathDevicePath = extern struct {
         type: DevicePathType,
         subtype: Subtype,
-        length: u16,
+        length: u16 align(1),
 
         pub fn getPath(self: *const FilePathDevicePath) [*:0]const u16 {
-            return @ptrCast([*:0]const u16, @alignCast(2, @ptrCast([*]const u8, self)) + @sizeOf(FilePathDevicePath));
+            return @ptrCast(
+                [*:0]const u16,
+                @alignCast(2, @ptrCast([*]const u8, self) + @sizeOf(FilePathDevicePath)),
+            );
         }
     };
```
If we're `alignCast`ing this, shouldn't the struct just be `align(2)`?

truemedian commented:
I finally found the important words for DevicePathProtocol.

> Therefore all nodes are byte-packed data structures that may appear on any byte boundary. All code
references to device path notes must assume all fields are unaligned.

Therefore, I've gone ahead and made all strings in device_path_protocol align(1) and removed all alignCasts.
I also added a note so that this doesn't get confusing in the future.

davidgm94 commented:
Is anything else required to merge this PR?

truemedian commented:
I think everything that broke has been fixed, lgtm

davidgm94 commented:
Maybe rebase against master to pass the CI?

truemedian commented:
As far as I can tell, the compiler is failing on the stdlib tests. From what I can tell this is because `std.os.uefi.Status` is a `enum(usize)` and most of its fields have the high bit set.

```
thread 93367 panic: integer cast truncated bits
/workspace/src/value.zig:1181:58: 0x66eda97 in toSignedInt (zig)
            .int_u64 => return @intCast(i64, self.castTag(.int_u64).?.data),
                                                         ^
codegen/llvm.zig:1508:64: 0x6706bdf in lowerDebugTypeImpl (zig)
codegen/llvm.zig:1444:36: 0x656715e in lowerDebugType (zig)
```

The code in `lowerDebugTypeImpl` points to <https://github.com/ziglang/zig/issues/645>.

It would seem that zig *cannot* generate debug info for these types, however I think the blind intCast to i64 is the wrong idea here, since it makes this enum impossible.

I believe that this error has never come up before since std.os.uefi has never been included in the test suite.

truemedian commented on lib/std/os/uefi.zig:
```diff
@@ -150,3 +150,8 @@ test "GUID formatting" {
     try std.testing.expect(std.mem.eql(u8, str, "32cb3c89-8080-427c-ba13-5049873bc287"));
 }
+
+comptime {
```
I've gone ahead and implemented both of these, it does however mean that uefi will continue to be untested on the CI. I can however confirm that all of the tests in uefi pass at the moment so we can get merge this PR and have a working std.os.uefi in 0.10 (at the least with debug info stripped).

andrewrk replied:
We can still add CI coverage for UEFI - let's just do it intentionally, via a new entry in the test matrix:

https://github.com/ziglang/zig/blob/20925b2f5c5c0ae20fdc0574e5d4e5740d17b4d6/test/tests.zig#L40

As you noted, I think it would be best to do that in a separate contribution so that we can get this particular PR landed in time for 0.10.0.

truemedian replied:
I did attempt to add uefi to the test targets, however a lot of the tests (behavior was the first that failed) that rely on the std.os interface (which uefi pretty much doesn't implement) were causing failures. Is there any way to get uefi to *only* run stdlib tests?

truemedian replied:
On a side note: I have *absolutely* no idea what is causing the compile error for debug on ci.ziglang.org

andrewrk commented on lib/std/os/uefi/protocols.zig:
```diff
@@ -43,3 +43,8 @@ pub usingnamespace @import("protocols/udp6_protocol.zig");
 pub usingnamespace @import("protocols/hii_database_protocol.zig");
 pub usingnamespace @import("protocols/hii_popup_protocol.zig");
+
+comptime {
```
there is another instance of this pattern here

truemedian replied:
Whoops, missed these

andrewrk commented:
Note that you can easily check locally whether the CI will fail by running this command:

```
stage3/bin/zig test ../lib/std/std.zig -target x86_64-linux
```

andrewrk reviewed (approved):
Ready to be merged when the CI tests pass.
